blocks/log.py

```diff
@@ -1,274 +1,418 @@
+class TrainingLog(Mapping):
+    r"""The log of training progress.
 
-    To support various backends (such as database tables) the
-    descendants of this class are expected to override the
-    `__getattr__`, `__setattr__`, `__hasattr__` methods.
+    A training log stores the training timeline, statistics and other
+    auxiliary information. Information is represented as nested dictionary
+    of iteration, key values.
 
-    Attributes
-    ----------
-    iterations_done : int
-        The number of iterations done.
-    epochs_done : int
-        The number of epochs done.
-    _epoch_ends : list
-        The numbers of the epochs last iterations.
+    In addition to the set of records of the training progress, a training
+    log has a status object whose attributes describe the current state of
+    training e.g. whether training has started, whether termination has
+    been requested, etc.
 
-    .. todo::
+    Parameters
+    ----------
+    hash_ : str, optional
+        A 12-byte hexidecimal string that identifies this experiment. If
+        given and a database backend is used, it will try to reload the
+        log. If not, a new hash will be created for this log.
+    backend : {'default', 'mongodb', 'sqlite'}, optional
+        The log can be stored in a variety of backends. The `default`
+        backend stores the values in a nested dictionary (a
+        :class:`DefaultOrderedDict` to be precise). This is a simple and
+        fast option, but the log is difficult to access for analysis during
+        training, and can only be saved as a pickled Python object.
+
+        The `mongo` option uses a MongoDB database server to store the
+        logs. The `sqlite` value will use a local SQLite file to store the
+        results. Both of these backends will allow you to analyze results
+        easily even during training, and will allow you to save multiple
+        experiments to the same database.
+    \*\*kwargs
+        Additional keyword arguments are passed to the log's backend e.g.
+        database configuration. See the relevant classes for details.
 
-        We need some notion of an attributes property. Examples of possible
-        properties include a docstring, a priority level to be used by a
-        printing extension to decide whether a certain attribute should be
-        printed or not.
+    Attributes
+    ----------
+    info : dict
+        A dictionary with information (metadata) about this experiment,
+        such as a description of the model, the package versions used, etc.
+        It also contains the unique hash that identifies experiments (e.g.
+        in the database).
+    status : :class:`TrainingStatus`
+        This dictionary contains information on the current status of the
+        training e.g. how many iterations/epochs have been completed,
+        whether training has finished, if errors occurred, etc.
+    hidden : dict
```

Comment:
@janchorowski , why should we specify the level every time we log altogether? This is redundant and we should rather do it once by a separate call like `add_annotation` from above. Also we might find it useful to annotate variables not only with the level but with something else, e.g. by the extension that produced them, so that one could later filter only values produced by a given extension.